File: SexyBiscuit.Engine/Scripting/ScriptComponent.cs
using Jint.Native;
using Jint.Runtime;
using SexyBiscuit.Engine.Core;

namespace SexyBiscuit.Engine.Scripting;

/// <summary>
/// A <see cref="Component"/> that attaches a JavaScript file to an <see cref="Actor"/>.
///
/// The JS file may define any or all of the following lifecycle functions:
/// <list type="bullet">
///   <item><c>onAwake()</c></item>
///   <item><c>onStart()</c></item>
///   <item><c>onUpdate(dt)</c></item>
///   <item><c>onFixedUpdate(dt)</c></item>
///   <item><c>onLateUpdate(dt)</c></item>
///   <item><c>onDestroy()</c></item>
///   <item><c>onCollisionEnter/Stay/Exit(data)</c> — data = <c>{ other, contactPoint:{x,y}, normal:{x,y}, relativeVelocity, tag, name }</c></item>
///   <item><c>onTriggerEnter/Stay/Exit(other)</c> — other = actor proxy object</item>
/// </list>
///
/// JavaScript errors are caught and reported through <see cref="ScriptDiagnostics"/>; they never crash the engine.
/// </summary>
public sealed class ScriptComponent : Component
{
    // -------------------------------------------------------------------------
    // Public state
    // -------------------------------------------------------------------------

    /// <summary>
    /// Path to the <c>.js</c> script file to run, relative to the project root.
    /// Example: <c>"Scripts/EnemyAI.js"</c>
    /// </summary>
    /// <remarks>
    /// Changing it on an attached component reloads the script immediately: the scene loader
    /// attaches first and sets properties second, and the editor's tools do the same, so a
    /// script that only ran when the path was set before attach never ran at all. When the actor
    /// is not yet in a scene the load waits for <see cref="Start"/>, so <c>onAwake</c> can use
    /// <c>Scene.*</c> — the order the browser runtime uses.
    /// </remarks>
    public string ScriptPath
    {
        get => _scriptPath;
        set
        {
            value ??= string.Empty;
            if (_scriptPath == value) return;
            _scriptPath = value;

            if (Actor == null || Actor.Scene == null) return;

            InitialiseRuntime();
            TryCall("onAwake");
            _awoken = true;
            if (_started) TryCall("onStart");
        }
    }
    private string _scriptPath = string.Empty;
    private bool   _started;
    private bool   _awoken;

    // invoke() calls made before the script loaded — a spawner configuring a script it has
    // just attached to an actor that is not yet in a scene. Replayed after onAwake, before onStart.
    private List<(string Function, JsValue[] Args)>? _pendingInvokes;

    /// <summary>
    /// The active Jint runtime for this component. Null until the script has loaded
    /// successfully. Replaced with a fresh instance on each hot-reload.
    /// </summary>
    public JintRuntime? Runtime { get; private set; }

    /// <summary>The last load error, or null when the script loaded. Surfaced by the editor's inspector.</summary>
    public string? Error { get; private set; }

    // -------------------------------------------------------------------------
    // Lifecycle — Component overrides
    // -------------------------------------------------------------------------

    public override void Awake()
    {
        // No path yet is the normal case for a component restored from a scene file; the
        // setter picks it up a moment later. No scene yet means the actor is still being
        // built, and Start is the first moment Scene.* can answer.
        if (string.IsNullOrWhiteSpace(_scriptPath) || Actor.Scene == null) return;

        InitialiseRuntime();
        TryCall("onAwake");
        _awoken = true;
    }

    public override void Start()
    {
        _started = true;

        if (Runtime == null && !string.IsNullOrWhiteSpace(_scriptPath))
        {
            InitialiseRuntime();
            if (Runtime != null && !_awoken)
            {
                TryCall("onAwake");
                _awoken = true;
            }
        }

        ReplayPendingInvokes();
        SubscribeToNetwork();
        TryCall("onStart");
    }

    public override void Update(float dt)
    {
        // A session may start after this component did -- a lobby script calls
        // Network.startServer from onUpdate -- so the subscription is retried until it
        // takes. It costs one null check per frame on a script that has no such hook.
        if (!_networkSubscribed) SubscribeToNetwork();
        TryCall("onUpdate", dt);
    }

    public override void FixedUpdate(float dt)
        => TryCall("onFixedUpdate", dt);

    public override void LateUpdate(float dt)
        => TryCall("onLateUpdate", dt);

    /// <summary>
    /// Subscribes to the message channel, but only for a script that declares the hook.
    /// </summary>
    /// <remarks>
    /// Subscribing unconditionally would put every scripted actor in the scene on a handler
    /// that does nothing, and a message-heavy game has hundreds. Called from Start, which is
    /// after the script has loaded and its functions are known.
    /// </remarks>
    private void SubscribeToNetwork()
    {
        if (_networkSubscribed || Runtime == null || !Runtime.HasFunction("onNetworkMessage")) return;

        var manager = Networking.NetworkManager.Instance;
        if (manager == null) return;

        _onNetworkMessage = (sender, type, payload) =>
        {
            if (Runtime == null) return;
            Runtime.CallFunctionWithJsArgs("onNetworkMessage",
                new Jint.Native.JsString(type),
                Runtime.Bridge.JsonToScript(payload),
                new Jint.Native.JsNumber(sender));
        };

        manager.OnMessage += _onNetworkMessage;
        _networkSubscribed = true;
    }

    private bool _networkSubscribed;
    private Action<int, string, System.Text.Json.Nodes.JsonNode?>? _onNetworkMessage;

    public override void OnDestroy()
    {
        if (_onNetworkMessage != null && Networking.NetworkManager.Instance is { } manager)
            manager.OnMessage -= _onNetworkMessage;
        _onNetworkMessage = null;
        _networkSubscribed = false;

        TryCall("onDestroy");

        // The script's own network handlers go with it. Without this a destroyed actor's
        // Network.on callback keeps firing into a dead Jint object every time a message
        // arrives, which is a leak that only shows up as a growing stall.
        Runtime?.Bridge.DisposeNetwork();
        Runtime?.Bridge.DisposeDarksGames();
    }

    public override void OnCollisionEnter(CollisionData data) => DispatchCollision("onCollisionEnter", data);
    public override void OnCollisionStay(CollisionData data)  => DispatchCollision("onCollisionStay",  data);
    public override void OnCollisionExit(CollisionData data)  => DispatchCollision("onCollisionExit",  data);
    public override void OnTriggerEnter(Actor other)          => DispatchTrigger("onTriggerEnter", other);
    public override void OnTriggerStay(Actor other)           => DispatchTrigger("onTriggerStay",  other);
    public override void OnTriggerExit(Actor other)           => DispatchTrigger("onTriggerExit",  other);

    // -------------------------------------------------------------------------
    // Cross-script calls
    // -------------------------------------------------------------------------

    /// <summary>
    /// Calls a top-level function the script defines and returns its result, or <c>undefined</c>
    /// when there is no such function or it throws. This is what another script reaches through
    /// <c>getComponent("ScriptComponent").invoke(name, ...args)</c>.
    /// </summary>
    public JsValue Invoke(string function, params JsValue[] args)
    {
        if (Runtime == null)
        {
            // Not loaded yet but about to be (the actor has no scene, so the load waits for
            // Start): keep the call, as the browser runtime does while it fetches the file.
            if (!string.IsNullOrWhiteSpace(_scriptPath) && Actor != null && Actor.Scene == null)
                (_pendingInvokes ??= new()).Add((function, args));
            return JsValue.Undefined;
        }

        try
        {
            return Runtime.Invoke(function, args);
        }
        catch (JavaScriptException ex)
        {
            LogScriptError(function, ex.Message);
        }
        catch (ExecutionCanceledException)
        {
            LogScriptError(function, "Execution cancelled — statement limit exceeded (100 000).");
        }
        catch (Exception ex)
        {
            LogScriptError(function, ex.Message);
        }

        return JsValue.Undefined;
    }

    // -------------------------------------------------------------------------
    // Hot-reload entry point
    // -------------------------------------------------------------------------

    /// <summary>
    /// Called by <see cref="ScriptHotReload"/> to reload the script in-place.
    /// Creates a fresh <see cref="JintRuntime"/> from the updated source file, then
    /// re-fires <c>onStart()</c> so the script can re-initialise its state.
    /// </summary>
    public void Reload()
    {
        InitialiseRuntime();

        // The component is already past Awake/Start — re-fire onStart so the
        // script can re-initialise any state that was set there.
        TryCall("onStart");
    }

    // -------------------------------------------------------------------------
    // Internals
    // -------------------------------------------------------------------------

    private void ReplayPendingInvokes()
    {
        if (_pendingInvokes == null) return;
        var pending = _pendingInvokes;
        _pendingInvokes = null;
        foreach (var (function, args) in pending) Invoke(function, args);
    }

    private void DispatchCollision(string hook, CollisionData data)
    {
        if (Runtime == null || !Runtime.HasFunction(hook)) return;

        try
        {
            Runtime.CallFunctionWithJsArgs(hook, Runtime.Bridge.WrapCollisionData(data));
        }
        catch (JavaScriptException ex)
        {
            LogScriptError(hook, ex.Message);
        }
        catch (Exception ex)
        {
            LogScriptError(hook, ex.Message);
        }
    }

    private void DispatchTrigger(string hook, Actor other)
    {
        if (Runtime == null || !Runtime.HasFunction(hook)) return;

        try
        {
            Runtime.CallFunctionWithJsArgs(hook, Runtime.Bridge.WrapActorAsProxy(other));
        }
        catch (JavaScriptException ex)
        {
            LogScriptError(hook, ex.Message);
        }
        catch (Exception ex)
        {
            LogScriptError(hook, ex.Message);
        }
    }

    private void InitialiseRuntime()
    {
        Error = null;

        if (string.IsNullOrWhiteSpace(ScriptPath))
        {
            Runtime = null;
            return;
        }

        // Read the JS source. A missing script is reported, never thrown.
        string source;
        try
        {
            source = File.ReadAllText(ProjectPaths.Resolve(ScriptPath));
        }
        catch (Exception ex)
        {
            Fail(null, $"could not read file — {ex.Message}");
            return;
        }

        // Construct a fresh runtime. JintRuntime creates its own Jint.Engine
        // and ScriptBridge, ensuring they always share the same engine instance.
        try
        {
            var runtime = new JintRuntime(Actor);
            runtime.Bridge.ScriptPath = ScriptPath;
            runtime.LoadScript(source);
            Runtime = runtime;
        }
        catch (JavaScriptException ex)
        {
            // Parse / execution error in the script itself.
            Fail("load", ex.Message);
        }
        catch (Exception ex)
        {
            Fail("load", ex.Message);
        }
    }

    private void Fail(string? hook, string message)
    {
        Runtime = null;
        Error   = message;
        ScriptDiagnostics.Report(ScriptDiagnosticLevel.Error, ScriptPath, hook, message);
    }

    /// <summary>
    /// Safely calls a lifecycle function by name, catching and reporting any
    /// JS or CLR exception without crashing the engine.
    /// </summary>
    private void TryCall(string functionName, params object[] args)
    {
        if (Runtime == null) return;

        try
        {
            Runtime.CallFunction(functionName, args);
        }
        catch (JavaScriptException ex)
        {
            LogScriptError(functionName, ex.Message);
        }
        catch (ExecutionCanceledException)
        {
            // MaxStatements limit was reached — report and continue.
            LogScriptError(functionName, "Execution cancelled — statement limit exceeded (100 000).");
        }
        catch (Exception ex)
        {
            LogScriptError(functionName, ex.Message);
        }
    }

    private void LogScriptError(string functionName, string message)
        => ScriptDiagnostics.Report(ScriptDiagnosticLevel.Error, ScriptPath, functionName, message);
}

File: html5/src/scripting/ScriptComponent.js
// -----------------------------------------------------------------------------
// ScriptComponent — runs a project `.js` file against one actor.
//
// Two script styles are supported from the same component:
//
//   * the flat-function style the C# engine's Jint bridge defines, where the
//     file declares `onStart`, `onUpdate` and friends at top level and reads the
//     `actor`, `transform`, `Input`, `Scene`, `Audio`, `Debug` and `Vector2`
//     globals — the form every existing project script is written in;
//
//   * an ES module exporting a Component subclass, which is the class-based API
//     this engine prefers for new work.
//
// The style is detected from the source, so a project can mix the two.
// -----------------------------------------------------------------------------

import { Component } from '../core/Component.js';
import { registerComponent } from '../core/TypeRegistry.js';
import { PropertyType as P } from '../core/PropertyTypes.js';
import { createScriptGlobals, wrapActor, wrapCollisionData, SCRIPT_HOOKS } from './ScriptBridge.js';

/** Attaches a JavaScript file to an actor. */
export class ScriptComponent extends Component {
    static schema = {
        scriptPath: { type: P.Asset, assetKind: 'script', default: '' },
    };

    constructor() {
        super();
        this._scriptPath = '';
        this._hooks = {};
        this._functions = {};       // every top-level function, for invoke()
        this._pendingInvokes = [];  // invoke() calls made before the script loaded
        this._delegate = null;      // a Component instance, for module-style scripts
        this._started = false;
        this._loading = null;
        this._initialised = false;

        /** The last load error, surfaced by the editor's inspector. */
        this.error = null;
    }

    /**
     * The project-relative path of the script.
     *
     * Assigning re-initialises immediately, because every loader attaches the
     * component first and sets its properties second.
     */
    get scriptPath() { return this._scriptPath; }
    set scriptPath(value) {
        const next = value == null ? '' : String(value);
        if (this._scriptPath === next) return;
        this._scriptPath = next;
        if (this.actor) this._initialise();
    }

    awake() { if (this._scriptPath) this._initialise(); }

    start() {
        this._started = true;

        // The scene loader attaches components and applies their properties
        // before the actor joins a scene, so when `scriptPath` was set there was
        // no asset manager to load through and initialisation was skipped. By
        // `start` the actor is in the scene, which is the first moment the
        // script can actually be fetched.
        if (this._scriptPath && !this._initialised) {
            this._initialise();
            return;   // onStart fires when the load resolves
        }

        // A script still loading gets its `onStart` once the load resolves.
        if (this._loading) return;
        this._call('onStart');
    }

    update(dt) { this._call('onUpdate', dt); }
    fixedUpdate(dt) { this._call('onFixedUpdate', dt); }
    lateUpdate(dt) { this._call('onLateUpdate', dt); }

    // A flat-function script gets the same wrapped shapes it would under Jint; a
    // module-style Component subclass gets the engine's own objects.
    onCollisionEnter(data) { this._call('onCollisionEnter', this._delegate ? data : wrapCollisionData(data)); }
    onCollisionStay(data) { this._call('onCollisionStay', this._delegate ? data : wrapCollisionData(data)); }
    onCollisionExit(data) { this._call('onCollisionExit', this._delegate ? data : wrapCollisionData(data)); }
    onTriggerEnter(other) { this._call('onTriggerEnter', this._delegate ? other : wrapActor(other)); }
    onTriggerStay(other) { this._call('onTriggerStay', this._delegate ? other : wrapActor(other)); }
    onTriggerExit(other) { this._call('onTriggerExit', this._delegate ? other : wrapActor(other)); }

    onDestroy() {
        this._call('onDestroy');
        if (this._delegate) {
            this._delegate.onDestroy?.();
            this._delegate = null;
        }
        this._hooks = {};
        this._functions = {};
    }

    /**
     * Calls a top-level function the script defines and returns its result — what
     * another script reaches through `getComponent("ScriptComponent").invoke(name, ...args)`.
     * Returns undefined when there is no such function or it throws.
     */
    invoke(name, ...args) {
        // A script is fetched, so the spawner that just attached it and wants to
        // configure it — `script.invoke("configure", damage)` — is early. The call
        // waits and runs once the script is in, after onAwake and before onStart,
        // which is the order the C# engine gives the same code.
        if (!this._initialised || this._loading) {
            this._pendingInvokes.push({ name, args });
            return undefined;
        }

        const fn = this._functions[name] ?? this._hooks[name]
            ?? (typeof this._delegate?.[name] === 'function' ? this._delegate[name].bind(this._delegate) : null);
        if (!fn) return undefined;
        try {
            return fn(...args);
        } catch (err) {
            console.error(`[Script Error] ${this._scriptPath} (${name}): ${err.message}`);
            return undefined;
        }
    }

    /** The same as invoke(). */
    call(name, ...args) { return this.invoke(name, ...args); }

    /** Reloads the script from source. Module-level state is lost, by design. */
    reload() {
        this._hooks = {};
        this._functions = {};
        this._delegate = null;
        this._initialised = false;
        this._initialise();
    }

    _initialise() {
        this._hooks = {};
        this._functions = {};
        this._delegate = null;
        this.error = null;

        const assets = this.actor?.scene?.engine?.assets;
        if (!assets) {
            // No asset manager yet: either the actor is not in a scene, which
            // `start` retries, or there is no host at all, which is a test or a
            // headless tool feeding source in through `setSource`.
            return;
        }

        this._initialised = true;
        this._loading = assets.loadText(this._scriptPath)
            .then((source) => {
                this._loading = null;
                this._instantiate(source);
                this._call('onAwake');
                this._replayPendingInvokes();
                if (this._started) this._call('onStart');
            })
            .catch((err) => {
                this._loading = null;
                this.error = err.message;
                console.error(`[ScriptComponent] ${this._scriptPath}: ${err.message}`);
            });
    }

    /** Compiles the source and captures whichever hooks it defines. */
    _instantiate(source) {
        if (isModuleSource(source)) {
            this._instantiateModule(source);
            return;
        }
        this._instantiateFunctions(source);
    }

    _instantiateFunctions(source) {
        const globals = createScriptGlobals(this.actor);
        const names = Object.keys(globals);

        // Each script gets its own function scope, so two components running the
        // same file keep separate state — the isolation Jint gets from one engine
        // per component.
        //
        // Every top-level function is collected, not only the hooks, so another
        // script can reach `takeDamage` through invoke(). The scan is deliberately
        // loose — a nested helper, or the word in a comment, is caught too — because
        // a name that is not a function at the top level collects as null and is
        // dropped; `typeof` on an undeclared name is 'undefined', not an error.
        const declared = new Set(SCRIPT_HOOKS);
        for (const match of source.matchAll(/\bfunction\s+([A-Za-z_$][\w$]*)\s*\(/g)) {
            declared.add(match[1]);
        }
        const collect = [...declared]
            .map((name) => `${JSON.stringify(name)}: typeof ${name} === 'function' ? ${name} : null`)
            .join(',\n    ');

        const body = `${source}\n\nreturn {\n    ${collect}\n};`;

        try {
            // eslint-disable-next-line no-new-func
            const factory = new Function(...names, body);
            const found = factory(...names.map((n) => globals[n]));
            for (const [name, fn] of Object.entries(found)) {
                if (!fn) continue;
                if (SCRIPT_HOOKS.includes(name)) this._hooks[name] = fn;
                this._functions[name] = fn;
            }
        } catch (err) {
            this.error = err.message;
            console.error(`[ScriptComponent] ${this._scriptPath} failed to compile: ${err.message}`);
        }
    }

    _instantiateModule(source) {
        // Module scripts are loaded by the asset manager, which caches the
        // evaluated module; this path only wires the exported class to the actor.
        const exported = source.__module ?? null;
        if (!exported) {
            this.error = 'Module scripts must be loaded through AssetManager.loadModule().';
            return;
        }

        const Ctor = exported.default ?? Object.values(exported).find(
            (v) => typeof v === 'function' && v.prototype instanceof Component);

        if (!Ctor) {
            this.error = 'Module script exports no Component subclass.';
            return;
        }

        this._delegate = new Ctor();
        this._delegate.actor = this.actor;
        this._delegate.awake?.();

        // Forward the whole lifecycle to the delegate.
        for (const [hook, method] of Object.entries(MODULE_HOOK_MAP)) {
            if (typeof this._delegate[method] === 'function') {
                this._hooks[hook] = (...args) => this._delegate[method](...args);
            }
        }
    }

    _call(hook, ...args) {
        const fn = this._hooks[hook];
        if (!fn) return;
        try {
            fn(...args);
        } catch (err) {
            console.error(`[Script Error] ${this._scriptPath} (${hook}): ${err.message}`);
        }
    }

    /**
     * Runs a script from source text without loading a file. Used by tests and
     * by the editor's console, and by any host that already has the text.
     */
    setSource(source) {
        this._hooks = {};
        this._functions = {};
        this._delegate = null;
        this._initialised = true;
        this._instantiate(source);
        this._call('onAwake');
        this._replayPendingInvokes();
        if (this._started) this._call('onStart');
    }

    _replayPendingInvokes() {
        const pending = this._pendingInvokes;
        this._pendingInvokes = [];
        for (const { name, args } of pending) this.invoke(name, ...args);
    }

    /** The hook names the loaded script actually defines. */
    get definedHooks() { return Object.keys(this._hooks); }
}
registerComponent(ScriptComponent, {
    category: 'Scripting',
    summary: 'Runs a project JavaScript file against this actor.',
});

const MODULE_HOOK_MAP = {
    onAwake: 'awake',
    onStart: 'start',
    onUpdate: 'update',
    onFixedUpdate: 'fixedUpdate',
    onLateUpdate: 'lateUpdate',
    onDestroy: 'onDestroy',
    onCollisionEnter: 'onCollisionEnter',
    onCollisionStay: 'onCollisionStay',
    onCollisionExit: 'onCollisionExit',
    onTriggerEnter: 'onTriggerEnter',
    onTriggerStay: 'onTriggerStay',
    onTriggerExit: 'onTriggerExit',
};

/** True when the source looks like an ES module rather than a flat script. */
function isModuleSource(source) {
    if (typeof source !== 'string') return true;   // already an evaluated module
    return /^\s*(export|import)\s/m.test(source);
}
